### ISSUE TITLE
getting kernel_client information from iopub/shell messages

### ISSUE THREAD

**Issue by jayendra13** [jupyter/jupyter_client#181]
I am creating a kernel_client instance here, which binds my call_back with the `iopub_stream`
```
    def register_kernel(self, access_token, language, callback):
        kernel_info = self.start_kernel(language)
        kernel_id, kernel_client, stream = kernel_info
        stream.on_recv_stream(callback)
        self.kernel_map[access_token] = {'kernel_id': kernel_id,
             'kernel_client': kernel_client,
             'iopub_stream': stream}

    def start_kernel(self, kernel_name):
    	kernel_id = self.km.start_kernel(kernel_name=kernel_name)
    	stream = self.km.connect_iopub(kernel_id)
    	cf = find_connection_file(kernel_id)
    	kernel_client = BlockingKernelClient(connection_file=cf)
    	kernel_client.load_connection_file()
    	kernel_client.start_channels()
    	return (kernel_id, kernel_client, stream)
```

` _on_zmq_reply` is my call_back
```
class WSHandler(websocket.WebSocketHandler):

	def open(self):
		self.write_message("WebSocket: Connected")

	#@handle_wsclosederror
	def _on_zmq_reply(self, m, msg_list):
		try:
			#print m
			#print type(m)
			#print msg_list
			for e in msg_list:
				print type(e)
				if type(e) == str:
					self.write_message(e)
		except websocket.WebSocketClosedError as err:
			self.kill_kernel()
```

If I have created multiple kernels,  on the call_back side I haven't a way to detect that which kernel_client corresponds to the current incoming message. 

I read the Ipython/Jupyter message format but that doesn't mention information about the identity of kernel in the messages.

**Comment by minrk** [comment]
Replies are associated with requests based on the `parent_header` on the reply. These are associates with two objects:

1. (primary) `parent_header['msg_id']` assocates it with a particular request message (identified via `request['header']['msg_id']`.
2. (secondary, less used) `parent_header['session']` associates it with a particular requesting client, in a Python KernelClient, this is `client.session.session`

But let's back up a bit: why do you want to look up the requesting client?

**Comment by jayendra13** [comment]
Let me introduce some background here. I am writing a web-app with tornado which has REPL frontend for two different languages(say matlab and python). On server side I am using jupyter_client to create two kernels for them. 

Now the catch is that the code in one REPL(say python), is dependent on the other one(matlab), as the matlab-REPL is writing some necessary files to disk for python. So I have to wait to execute the python code until matlab code finishes. I don't think that the tornado futures can help me here, so I have to explicitly track all kernel_iopub messages to maintain the order of execution.

**Comment by jayendra13** [comment]
or do you think is there a way to do this asynchronously?
How to execute line 29 only after 27 has beed executed, does tornado.concurrent.Future will help here?
```python
1  from zmq.eventloop import ioloop
2  ioloop.install()
3  from jupyter_client import BlockingKernelClient, find_connection_file
4  from jupyter_client import MultiKernelManager
5
6  km = MultiKernelManager()
7
8  def callback(m ,msg_list):
9     for m_ in msg_list:
10         print m_
11
12 def start_kernel():
13    kernel_id = km.start_kernel()
14    stream = km.connect_iopub(kernel_id)
15    cf = find_connection_file(kernel_id)
16    kernel_client = BlockingKernelClient(connection_file=cf)
17    kernel_client.load_connection_file()
18    kernel_client.start_channels()
19    return kernel_client, stream
20
21 k1, s1 = start_kernel()
22 s1.on_recv_stream(callback)
23
24 k2, s2 = start_kernel()
25 s2.on_recv_stream(callback)
26
27 k1.execute("import time; time.sleep(10)")
28
29 k2.execute("print 'Hello'")
```

**Comment by minrk** [comment]
You can use Futures to express the ordering, but jupyter-client doesn't give you the Futures you want, so you would need to build them yourself.

Here's a version that uses Futures to chain the two executions:


```python
from pprint import pprint
from functools import partial

from tornado.concurrent import Future
from tornado import gen

from zmq.eventloop import ioloop
from zmq.eventloop.zmqstream import ZMQStream
ioloop.install()

from jupyter_client import MultiKernelManager

reply_futures = {}

km = MultiKernelManager()

def reply_callback(session, stream, msg_list):
    idents, msg_parts = session.feed_identities(msg_list)
    reply = session.deserialize(msg_parts)
    parent_id = reply['parent_header'].get('msg_id')
    reply_future = reply_futures.get(parent_id)
    if reply_future:
        reply_future.set_result(reply)

def start_kernel():
    kernel_id = km.start_kernel()
    kernel_client = km.get_kernel(kernel_id).client()
    kernel_client.start_channels()
    shell_stream = ZMQStream(kernel_client.shell_channel.socket)
    return kernel_client, shell_stream

k1, s1 = start_kernel()
s1.on_recv_stream(partial(reply_callback, k1.session))

k2, s2 = start_kernel()
s2.on_recv_stream(partial(reply_callback, k2.session))


@gen.coroutine
def k1_then_k2(code1, code2):
    """Run code1 on k1, wait for it to finish, then run code2 on k2"""
    msg_id = k1.execute(code1)
    f = reply_futures[msg_id] = Future()
    reply = yield f # wait for reply
    pprint(reply)

    msg_id2 = k2.execute(code2)
    f2 = reply_futures[msg_id2] = Future()
    reply = yield f2 # wait for second reply
    pprint(reply)

ioloop.IOLoop.current().run_sync(lambda : k1_then_k2(
    "import time; time.sleep(10)",
    "print('Hello')",
))
```

**Comment by jayendra13** [comment]
Thanks @minrk this example greatly helped me.

I thing which I still don't understand is that the signature of the callback function `reply_callback`. I know that when messages arrives in the buffer of shell stream callback function get executed, but I don't exactly which module/function calls it? and with what parameters?

**Comment by minrk** [comment]
`s1.on_recv_stream(callback)` means that whenever a message arrives on `s1`, it will call:

```python
callback(s1, msg_list)
```

where `msg_list` is the multi-part zmq message (a list of bytes objects or zmq.Frame objects).

In this example, I needed to associate the `session` object for each client with the messages it receives in order to do the deserialization. I used [`functools.partial`](https://docs.python.org/3/library/functools.html#functools.partial) to do this. `partial` lets me take a callback with the signature `callback(session, stream, msg_list)` and get a *new* callback with the signature `new_callback(stream, msg_list)` that will call the original with the extra `session` argument first.
